### PR TITLE
Add caching to CI's and fix docs

### PR DESCRIPTION
**Pull request by cbaker6**
This PR does the following:

- [x] Adds caching to circle ci & travis
- [x] Update travis and use for deployments only
- [x] Fix docs deployment

### CODE REVIEW

**Comment by cbaker6** [comment]
Is there a reason why travis and circle-ci are setup to do similar things? It looks like most of travis was migrated to circle ci, but the travis might just be left over?

**Comment by drdaz** [comment]
> Is there a reason why travis and circle-ci are setup to do similar (e.g. jazzy and cocoapods) things? It looks like most of travis was migrated to circle ci, but the travis might just be left over?

Not sure which way the move was going actually - to travis or to circleci. I added the circleci deployment + cocoapods + carthage targets to prs at some point because we were merging things that caused those targets to break when we merged. So we weren't always seeing the results from the Travis tests. 

Whether that was 'just' presentation or they really weren't being run concerned me less than the fact that we were blindly making master unstable.

**Comment by cbaker6** [comment]
@drdaz I see... it looks like from your test fix PR, you have everything working in circle? 

How many tests does circle run in parallel? For travis, it looks like OS X allows two to be run at a time.

I don't know much about circle. I do know travis has rake preinstalled, so technically the only thing bundler needs to install in travis is jazzy and it's dependencies

If you want, it might not take me that long to get the tests running in travis, then you can compare speed/feedback of travis and circle and decide which one is better for this sdk

**Comment by cbaker6** [comment]
Actually, looking at the test times in your PR, Circle seems to be way faster, but has longer queue times, might be best to leave it all there and remove travis?

**Comment by cbaker6** [comment]
This should be ready for review...

Are the pod and carthage deployments working? It doesn't look like pod is using pod lint in circle

**Comment by cbaker6** [comment]
@TomWFox can you setup the token for this branch the way you did for the swift one? If so, I can fix the doc deployment in this PR

**Comment by TomWFox** [comment]
done

**Comment by cbaker6** [comment]
@drdaz @TomWFox I think you may need to try the latest update on a branch to see if it deploys the docs correctly. I don't think I can get it to deploy from a PR

**Comment by cbaker6** [comment]
@drdaz to speed these up, I’m removing “deployment” and “cocoapods” from circle since they are running in Travis. Basically there’s some double work which is adding to the build times.

Also, the caching in circle seems to save 3-4 minutes per workflow item, so that should save ~15 additional minutes in the build. I’m sure I setup the Travis cache correctly, and I’m guessing we will see some time saving there too once the PR is merged.

Let me know if you see any issues with the setup above.

I’ll push my changes after your master build finishes as I’m not sure how circle prioritizes it’s queue

**Comment by cbaker6** [comment]
@drdaz has this error popped up randomly before?

```

Failing tests:
	ParseUnitTests-iOS:
		-[FieldOperationDecoderTests testDecodingAddOperations]
```

Not sure if it’s random or occurred because I bumped to Xcode 11.6

**Comment by drdaz** [comment]
> @drdaz to speed these up, I’m removing “deployment” and “cocoapods” from circle since they are running in Travis (Should save ~45 minutes total). Basically there’s some double work which is adding to the build times.
> 
> Leveraging Travis and Circle seems to be a better setup for this project as opposed to using one over the other:
> 
> Travis
> 
> * Deployment
> * Jazzy docs
> * Cocoapods
> 
> Circle
> 
> * iOS
> * macOS
> * parseUI
> * twitterUtils
> * facebookUtils
> * carthage
> 
> Also, the caching in circle seems to save 3-4 minutes per workflow item, so that should save ~15 additional minutes in the build. I’m sure I setup the Travis cache correctly, and I’m guessing we will see some time saving there too once the PR is merged.
> 
> Let me know if you see any issues with the setup above.

The only issue that comes to mind is that I *believe* we've seen the travis build not always run. I added the deploy/carthage/cocoapods circleci builds at some point because we merged some changes that broke those builds on master where the PR didn't show there was anything wrong.

**Comment by drdaz** [comment]
> @drdaz has this error popped up randomly before?
> 
> ```
> 
> Failing tests:
> 	ParseUnitTests-iOS:
> 		-[FieldOperationDecoderTests testDecodingAddOperations]
> ```
> 
> Not sure if it’s random or occurred because I bumped to Xcode 11.6

Yeah I've seen this before. In fact it occurred on the first test run after I merged my test fixes branch - did a sly re-run to get the green bar.

You almost certainly didn't cause this 🙂

**Comment by cbaker6** [comment]
> The only issue that comes to mind is that I believe we've seen the travis build not always run. I added the deploy/carthage/cocoapods circleci builds at some point because we merged some changes that broke those builds on master where the PR didn't show there was anything wrong.

How would you like me to move forward? I made some tweaks to travis and if it's broken after the merge, I can do the fixes and help maintain it. Having everything on circle will make for a slow build. I don't know much about rake but it didn't seem circle was deploying the pods

**Comment by drdaz** [comment]
Let's try yanking the CircleCI targets. We can watch out for travis not running and act if it happens again.

**Comment by cbaker6** [comment]
So the times seem to be around the following (note that travis and circle ran in parallel and all scenarios, so the bottleneck was the longest running):

- Now: Circle (~52 minutes with cache). Before: ~2 hours without cache, with cocoapods and deployment
- Now: Travis (~33 minutes, when this is merged I think it will take advantage of cache). The difference here is there are now 3 builds instead of 2 to prevent timeouts. Before: same time as now
 
So now the builds finish in ~52 minutes as opposed to ~2 hours

**Comment by cbaker6** [comment]
@drdaz it looks like this failure popped up again:

```
ParseUnitTests-macOS:
		-[PinningObjectStoreTests testUnpinAllObjects]
```

File: .travis.yml [hunk 1]
```diff
@@ -1,54 +1,59 @@
...
+      script:
+        - ./Scripts/jazzy.sh
+      deploy:
         - provider: pages
-          cleanup: false
-          token:
-            secure: m7YayvmctH5lyHj13EV2j898P5xfEfN/ZvFm2gL/xveTNOGaET1O4ZmGmHApxC2QwIZvtwmO3U9pmB4XVVQ7kIlTsl01h84y8gx1XQqEKoQINcaVaNCsh9EdxAqj08aiwe/1dp8W8rI/ptOrxzWJDWcxKx5tMIbBzctGeL3GDKphVvD44s43oosqQCKzP+qHiI6rRRxu+0QccdbTagwd0X/yaLZ0IbTo35hWFEyJ6NfJnmCW/lNwWj86YksLCZx0eKdxUSQlIhnm0upTyowudt9bv8vSdEYi2AtRKoana/X53593O8RAVEc+S6NgMmeQ1/ZD1eDeyYLzN/z21EHb4yiwTnMUUhDBMTn8KKTiiSVDQOfAyRwtKF+1qN2iDr9Dh3A8+dCxOxJJpIPImZimt8KMCKB9WTtnJhvIoMCDpc8l29XM/oBxjCiBFxPClb9CZ74T5b+xl+yIqy2u8RiaZ1Va5vmWjMSXDsGHoZ0rJTOBkBBhyr+47yIw2VeY2z2q5hiALV3DnE5Esxyddhw7OXNZcvKriujnxcTOJwphm4XBMI/3rSH27uz7AaNdUHfTwaKqimPR06Kdv65hKmFuA5p/GuRdSYfYASvXYYrPjqKkst1FP6sWjJtpXvZIT/06RXDbGXZ6LB3TFHpOahPy8ID746OyBKgWzOB3ZCNmFCk=
-          local_dir: docs/
+          skip_cleanup: true
+          github_token: $PERSONAL_ACCESS_TOKEN
```
Comment (TomWFox) [inline .travis.yml:40]: This is a to-do for me - the token is not a secure var in Travis yet, it needs to be added for this to work

**Comment by drdaz** [comment]
> @drdaz it looks like this failure popped up again:
> 
> ```
> ParseUnitTests-macOS:
> 		-[PinningObjectStoreTests testUnpinAllObjects]
> ```

The one that used to come up regularly was testFetchPin.

I'd lean towards this being related to OCMock... the current release has some issues with multithreading that can cause test crashes. I'm guessing this can happen literally anywhere, since I've seen it happen on one of our tests that was just testing a simple constructor returned an object.

File: .travis.yml (hunk 1 above)
Comment (TomWFox) [reply .travis.yml]: I just changed the environment var as there was already one in travis.

**Review by drdaz — APPROVED** [review]
LGTM!

**Comment by drdaz** [comment]
@TomWFox I think merging is blocked for your approval on the changes you requested. I can't merge at least.

**Pull request merged**